docs/devs/e2e-tests.md

```diff
@@ -0,0 +1,189 @@
+
+### Testing actions and effects
+
+While Ozone relies on its components default behaviour and features, we do encompass in our E2E test suite the actions performed when using the user experience of the components and the effects observed throught the user experience of those components. This ensures that test are truly end-to-end with a focus on the end-user experience.
+
+To write a test case:
+
+- Identify the user interactions and functionalities to test between a pair of HIS components.
+- Use Playwright’s API to script the actions within each test scenario, such as navigating pages, clicking buttons, filling forms, and verifying UI elements.
+
+### Utilizing Playwright
+
+Leverage Playwright’s comprehensive API to engage with the browser:
+
+- Navigate to pages.
+- Interact with page elements.
+- Execute actions.
+- Assert expected behaviors using methods like `expect`, `toEqual`, `toContain`, etc.
+
+For detailed usage, refer to [Playwright's documentation <small>:fontawesome-solid-arrow-up-right-from-square:</small>](https://playwright.dev/docs/intro).
+
+Once you've written your test, use the appropriate test runner to check that the test works as expected. For instance for the OpenMRS-SENAITE flows written in `openmrs-senaite-flows.spec.js`, run:
+```
+npx playwright test openmrs-senaite-flows
+```
+
+### Analyzing a sample test case
+
+Again, in the context of the data flows between OpenMRS and SENAITE, let us verify that ordering lab test for an OpenMRS patient does create the corresponding SENAITE client with an analysis request. Let us look at the sample test case code below:
+
+```javascript
+import { test, expect } from '@playwright/test';
+import { HomePage } from '../utils/functions/testBase';
+import { patientName } from '../utils/functions/testBase';
+
+let homePage: HomePage;
+
+test.beforeEach(async ({ page }) => {
+  homePage = new HomePage(page);
+  await homePage.initiateLogin();
+  await expect(page).toHaveURL(/.*home/);
+  await homePage.createPatient();
+  await homePage.startPatientVisit();
+});
+
+test('Ordering a lab test for an OpenMRS patient creates the corresponding SENAITE client with an analysis request.', async ({ page }) => {
+  // setup
+  homePage = new HomePage(page);
+  await homePage.goToLabOrderForm();
+  await page.click('button:has-text("Add")');
+  await page.selectOption('#tab select', '857AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA');
+  await homePage.saveLabOrder();
+
+  // replay
+  await homePage.goToSENAITE();
+  await expect(page).toHaveURL(/.*senaite/);
+
+  // verify
+  await homePage.searchClientInSENAITE();
+  const clientName = `${patientName.firstName} ${patientName.givenName}`;
+  const client = await page.$('table tbody tr:nth-child(1) td.contentcell.title div span a:has-text("' + clientName + '")');
+  await expect(client).toBeVisible();
```

Comment:
@kdaud @Ruhanga I think the sample test case does not illustrate well what each test case should be about. Namely "performing actions in component A and verifying expected outcome(s) in component B".

The "replay" here is ordering a lab test, the "verify" here is to observe that the corresponding client with an analysis request is in SENAITE as it should.

Reply:
Correct @mks-d. Thanks for the catch. I guess this was missed through some refactor as most of the other tests reflect the correct arrangement. For example this one:

```javascript
test('Editing the details of an OpenMRS patient with a synced lab order edits the corresponding SENAITE client details.', async ({ page }) => {
  // setup
  homePage = new HomePage(page);
  await homePage.goToLabOrderForm();
  await page.getByRole('button', { name: 'Add', exact: true }).click();
  await page.locator('#tab select').selectOption('857AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA');
  await homePage.saveLabOrder();
  await homePage.goToSENAITE();
  await expect(page).toHaveURL(/.*senaite/);
  await homePage.searchClientInSENAITE();
  const client = await page.locator('table tbody tr:nth-child(1) td.contentcell.title div span a');
  await expect(client).toContainText(`${patientName.firstName + ' ' + patientName.givenName}`);

  // replay
  await page.goto(`${O3_URL}`);
  await homePage.searchPatient(`${patientName.firstName + ' ' + patientName.givenName}`);
  await homePage.updatePatientDetails();

  // verify
  await homePage.goToSENAITE();
  await homePage.searchClientInSENAITE();

  await expect(client).toContainText(`${patientName.updatedFirstName}` + ' ' + `${patientName.givenName }`);
});

```

Reply:
Sure, there are few tests that needed a refactor.  I've logged OZ-561 to update those tests to align with the structure of Given When and Then.